### PR TITLE
Update: Typescript navi directory

### PR DESCRIPTION
**Pull request by kevinhinterlong**
## Description
We should start adopting typescript, this is somewhat of a test run (navi-directory is a small/easy to deal with addon)

Notable restriction from [typescript in ember](https://github.com/typed-ember/ember-cli-typescript#account-for-addon-build-pipeline-changes)
- The addon's `app/` folder should remain as `.js`
- Apps must use `.js` to override addon defaults (see [this](https://github.com/yahoo/navi/blob/dc9a507b0bbf6b42534e322207ef0b230d3152c1/packages/directory/tests/dummy/app/services/directories.js) from dummy app)

## Proposed Changes
- add typescript support
- add more eslint plugins
- add basic `Dict<T>` (`{string: T}`) object, and `TODO<T = any>` type (stolen from [emberclear](https://github.com/NullVoxPopuli/emberclear/blob/648a85f14c180cc508592ec4913ee24bf1f24980/packages/frontend/types/index.d.ts#L43-L44))

## License

I confirm that this contribution is made under an MIT license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

**Comment by kevinhinterlong** [comment]
<details>
<summary><code>ember ts:precompile</code> currently has some errors</summary>

```sh
addon/controllers/directory.ts:97:28 - error TS2345: Argument of type '{ queryParams: object; }' is not assignable to parameter of type 'string'.

97     this.transitionToRoute({ queryParams });
                              ~~~~~~~~~~~~~~~

addon/controllers/directory/my-data.ts:11:34 - error TS7016: Could not find a declaration file for module 'dummy/routes/directory/my-data'. '/Users/hinterlong/Desktop/dev/navi/packages/directory/app/routes/directory/my-data.js' implicitly has an 'any' type.

11 import DirectoryMyDataRoute from 'dummy/routes/directory/my-data';
                                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

addon/controllers/directory/my-data.ts:51:28 - error TS2345: Argument of type '{ queryParams: object; }' is not assignable to parameter of type 'string'.

51     this.transitionToRoute({ queryParams });
                              ~~~~~~~~~~~~~~~


Found 3 errors.
```
</details>

Blocked on https://github.com/DefinitelyTyped/DefinitelyTyped/pull/42808

File: packages/directory/addon/controllers/directory.ts [hunk 1]
```diff
@@ -58,42 +51,56 @@ export default class DirectoryController extends Controller {
...
   }
 
   /**
    * @action updateQueryParams - update to the new query params
-   * @param {Object} queryParams
+   * @param {object} queryParams
    */
   @action
-  updateQueryParams(queryParams) {
+  updateQueryParams(queryParams: object) {
```
Comment (cythrawll) [inline packages/directory/addon/controllers/directory.ts]: `{ [key: string]: string; }`

Comment (kevinhinterlong) [reply packages/directory/addon/controllers/directory.ts]: I think query params can actually be `boolean | number | array | string`

[ember queryparams guides](https://guides.emberjs.com/release/routing/query-params/)
[current type definitions](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/80fef0454291a5c8cd70a4fcaae767b4da60314f/types/ember__controller/index.d.ts#L13)

But we could add a type for DirectoryQueryParams explicitly

```ts
type DirectoryQueryParams = {
  filter: string | undefined,
  type: string | undefined,
  sortBy: string,
  sortDir: string,
  q: string
}
```

File: packages/directory/package.json [hunk 2]
```diff
@@ -18,12 +18,14 @@
   "scripts": {
     "build": "ember build --environment=production",
     "lint:hbs": "ember-template-lint .",
-    "lint:js": "eslint .",
+    "lint:js": "eslint . && tsc --noEmit",
+    "lint": "npm run lint:hbs && npm run lint:js",
     "start": "ember serve",
-    "test": "export COVERAGE=true; ember test --test-port 0"
+    "test": "export COVERAGE=true; ember test --test-port 0",
+    "prepublishOnly": "ember ts:precompile",
```
Comment (jkusa) [inline packages/directory/package.json]: Have you executed `ember ts:precompile` yet? I had to change `test-helpers.js` to `test-heplers.ts` for `qunit-dom` to work.

File: packages/directory/addon/routes/directory/my-data.ts [hunk 3]
```diff
@@ -44,16 +44,14 @@ export default class MyDataRoute extends Route {
...
-   * @param {Object} queryParams - all directory query params
+   * @param {object} user
+   * @param {object} queryParams - all directory query params
    */
-  async _fetchItems(user, { type, filter }) {
-    let reports,
-      dashboards,
-      items = arr();
+  async _fetchItems(user: TODO, { type, filter }: DirectoryParams) {
+    let items = arr();
```
Comment (jkusa) [inline packages/directory/addon/routes/directory/my-data.ts]:
```suggestion
    const items = arr();
```

File: packages/directory/addon/services/directories.ts [hunk 4]
```diff
@@ -5,12 +5,26 @@
...
+}
+
 export default class DirectoriesService extends Service {
   /**
-   * @property {Array} directories - Object list with info on each directory
+   * @property {Array<Directory>} directories - Object list with info on each directory
    */
-  @computed
-  get directories() {
+  @computed()
```
Comment (jkusa) [inline packages/directory/addon/services/directories.ts]: Does this need to be a `getter`?

File: packages/directory/addon/utils/type-utils.ts [hunk 5]
```diff
@@ -0,0 +1,11 @@
+import Route from '@ember/routing/route';
```
Comment (jkusa) [inline packages/directory/addon/utils/type-utils.ts]:
```suggestion
/**
 * Copyright 2020, Yahoo Holdings Inc.
 * Licensed under the terms of the MIT license. See accompanying LICENSE.md file for terms.
 */
 
import Route from '@ember/routing/route';
```

File: packages/directory/tests/dummy/app/services/directories.js [hunk 6]
```diff
@@ -12,7 +15,7 @@ export default class Directories extends DirectoryService {
           {
             name: 'Favorites',
             icon: 'star-o',
-            queryParam: { filter: 'favorites' }
+            queryParams: { filter: 'favorites' }
```
Comment (jkusa) [inline packages/directory/tests/dummy/app/services/directories.js]: non-getter?

**Pull request merged**